### ISSUE TITLE
Training pipeline v2 concerns

### ISSUE THREAD

**Issue by gcp** [gcp/leela-zero#1061]
I'm very concerned that the v2 training pipeline is still broken in some ways.

Network training with the old code:

```
step 536000, policy=1.70719 training accuracy=60.0654%, mse=0.189426                                    
Model saved in file: /home/morbo/git/leela-zero/training/tf/leelaz-model-536000
Leela weights saved to /home/morbo/git/leela-zero/training/tf/leelaz-model-536000.txt
step 537000, policy=1.68706 mse=0.189527 reg=0.100594 total=2.54577 (245.861 pos/s)
step 538000, policy=1.69586 mse=0.188917 reg=0.100576 total=2.5521 (312.572 pos/s)
step 539000, policy=1.70284 mse=0.187572 reg=0.100557 total=2.55369 (312.547 pos/s)
step 540000, policy=1.69672 mse=0.18866 reg=0.100539 total=2.5519 (312.557 pos/s)
step 541000, policy=1.69857 mse=0.188329 reg=0.10052 total=2.55241 (312.537 pos/s)
step 542000, policy=1.69134 mse=0.188932 reg=0.100501 total=2.54757 (312.515 pos/s)
step 543000, policy=1.69543 mse=0.187976 reg=0.100483 total=2.54782 (312.469 pos/s)
step 544000, policy=1.69255 mse=0.188601 reg=0.100464 total=2.54742 (312.546 pos/s)
step 544000, policy=1.68782 training accuracy=60.6523%, mse=0.193161
Model saved in file: /home/morbo/git/leela-zero/training/tf/leelaz-model-544000
Leela weights saved to /home/morbo/git/leela-zero/training/tf/leelaz-model-544000.txt
step 545000, policy=1.69817 mse=0.188129 reg=0.100447 total=2.55113 (245.83 pos/s)
step 546000, policy=1.69537 mse=0.187959 reg=0.100428 total=2.54763 (312.531 pos/s)
step 547000, policy=1.69902 mse=0.188232 reg=0.10041 total=2.55236 (312.555 pos/s)
step 548000, policy=1.69523 mse=0.188293 reg=0.100392 total=2.54879 (312.552 pos/s)
step 549000, policy=1.68944 mse=0.189317 reg=0.100374 total=2.54708 (312.571 pos/s)
step 550000, policy=1.6936 mse=0.188422 reg=0.100355 total=2.54764 (312.593 pos/s)
step 551000, policy=1.69238 mse=0.188502 reg=0.100337 total=2.54673 (312.514 pos/s)
step 552000, policy=1.69912 mse=0.188462 reg=0.100319 total=2.55329 (312.514 pos/s)
step 552000, policy=1.70371 training accuracy=60.5039%, mse=0.19109
Model saved in file: /home/morbo/git/leela-zero/training/tf/leelaz-model-552000
Leela weights saved to /home/morbo/git/leela-zero/training/tf/leelaz-model-552000.txt
step 553000, policy=1.6984 mse=0.18876 reg=0.100301 total=2.55374 (245.905 pos/s)
step 554000, policy=1.6998 mse=0.187627 reg=0.100282 total=2.55059 (312.567 pos/s)
step 555000, policy=1.69533 mse=0.187647 reg=0.100264 total=2.54619 (312.543 pos/s)
step 556000, policy=1.69995 mse=0.189174 reg=0.100245 total=2.55689 (312.554 pos/s)
step 557000, policy=1.69554 mse=0.188209 reg=0.100227 total=2.5486 (312.575 pos/s)
step 558000, policy=1.70231 mse=0.188391 reg=0.100208 total=2.55608 (312.555 pos/s)
step 559000, policy=1.70403 mse=0.187681 reg=0.10019 total=2.55494 (312.533 pos/s)
step 560000, policy=1.69426 mse=0.187946 reg=0.100172 total=2.54622 (312.484 pos/s)
step 560000, policy=1.70592 training accuracy=60.1689%, mse=0.187416
Model saved in file: /home/morbo/git/leela-zero/training/tf/leelaz-model-560000
Leela weights saved to /home/morbo/git/leela-zero/training/tf/leelaz-model-560000.txt
step 561000, policy=1.69498 mse=0.18807 reg=0.100155 total=2.54742 (245.849 pos/s)
step 562000, policy=1.68689 mse=0.188496 reg=0.100137 total=2.54101 (312.547 pos/s)
step 563000, policy=1.69464 mse=0.189104 reg=0.100119 total=2.55118 (312.554 pos/s)
step 564000, policy=1.69849 mse=0.189038 reg=0.100101 total=2.55475 (312.534 pos/s)
step 565000, policy=1.70006 mse=0.189126 reg=0.100083 total=2.55665 (312.537 pos/s)
step 566000, policy=1.69706 mse=0.187191 reg=0.100063 total=2.54588 (312.51 pos/s)
step 567000, policy=1.70139 mse=0.188182 reg=0.100046 total=2.55416 (312.565 pos/s)
step 568000, policy=1.69752 mse=0.188427 reg=0.100028 total=2.55125 (312.551 pos/s)
step 568000, policy=1.69253 training accuracy=60.5889%, mse=0.188714
Model saved in file: /home/morbo/git/leela-zero/training/tf/leelaz-model-568000
Leela weights saved to /home/morbo/git/leela-zero/training/tf/leelaz-model-568000.txt
step 569000, policy=1.69926 mse=0.187446 reg=0.10001 total=2.54905 (245.805 pos/s)
step 570000, policy=1.69814 mse=0.188169 reg=0.0999929 total=2.55081 (312.519 pos/s)
step 571000, policy=1.69845 mse=0.188102 reg=0.0999755 total=2.55084 (312.537 pos/s)
step 572000, policy=1.69701 mse=0.189579 reg=0.0999572 total=2.55529 (312.555 pos/s)
```

Same network, continuing training, but using v2 code:

```
Model saved in file: /home/morbo/git/leela-zero/training/tf/leelaz-model-576000
Leela weights saved to /home/morbo/git/leela-zero/training/tf/leelaz-model-576000.txt
step 577000, policy=1.71688 mse=0.192603 reg=0.100024 total=2.58732 (236.009 pos/s)
step 578000, policy=1.7155 mse=0.191874 reg=0.100024 total=2.58302 (300.619 pos/s)
step 579000, policy=1.71673 mse=0.192965 reg=0.100024 total=2.58861 (300.564 pos/s)
step 580000, policy=1.71172 mse=0.193035 reg=0.100022 total=2.58389 (300.537 pos/s)
step 581000, policy=1.71498 mse=0.193549 reg=0.10002 total=2.5892 (300.539 pos/s)
step 582000, policy=1.71038 mse=0.191926 reg=0.10002 total=2.57811 (300.596 pos/s)
step 583000, policy=1.72087 mse=0.194064 reg=0.10002 total=2.59714 (300.669 pos/s)
step 584000, policy=1.71855 mse=0.192677 reg=0.100019 total=2.58928 (300.483 pos/s)
step 584000, policy=1.70126 training accuracy=60.0117%, mse=0.193385
Model saved in file: /home/morbo/git/leela-zero/training/tf/leelaz-model-584000
Leela weights saved to /home/morbo/git/leela-zero/training/tf/leelaz-model-584000.txt
step 585000, policy=1.71406 mse=0.192978 reg=0.100018 total=2.58599 (236.578 pos/s)
step 586000, policy=1.72027 mse=0.19223 reg=0.100015 total=2.58921 (300.569 pos/s)
step 587000, policy=1.71467 mse=0.192039 reg=0.100016 total=2.58284 (300.407 pos/s)
step 588000, policy=1.7188 mse=0.192483 reg=0.100013 total=2.58874 (300.518 pos/s)
step 589000, policy=1.72077 mse=0.192778 reg=0.100011 total=2.59189 (300.632 pos/s)
step 590000, policy=1.71786 mse=0.192563 reg=0.100009 total=2.58812 (300.513 pos/s)
step 591000, policy=1.71263 mse=0.193193 reg=0.100008 total=2.58541 (300.532 pos/s)
step 592000, policy=1.72076 mse=0.192684 reg=0.100007 total=2.5915 (300.444 pos/s)
step 592000, policy=1.70599 training accuracy=59.1592%, mse=0.220706
Model saved in file: /home/morbo/git/leela-zero/training/tf/leelaz-model-592000
Leela weights saved to /home/morbo/git/leela-zero/training/tf/leelaz-model-592000.txt
step 593000, policy=1.71129 mse=0.192797 reg=0.100005 total=2.58248 (236.578 pos/s)
step 594000, policy=1.71302 mse=0.19448 reg=0.100002 total=2.59094 (300.626 pos/s)
step 595000, policy=1.71482 mse=0.19287 reg=0.100001 total=2.5863 (300.584 pos/s)
step 596000, policy=1.71936 mse=0.192509 reg=0.0999977 total=2.58939 (300.521 pos/s)
step 597000, policy=1.72781 mse=0.192803 reg=0.0999957 total=2.59902 (300.666 pos/s)
step 598000, policy=1.72271 mse=0.192713 reg=0.0999936 total=2.59356 (300.541 pos/s)
step 599000, policy=1.72014 mse=0.192526 reg=0.0999926 total=2.59024 (300.558 pos/s)
step 600000, policy=1.72208 mse=0.193163 reg=0.09999 total=2.59472 (300.652 pos/s)
step 600000, policy=1.67407 training accuracy=60.3906%, mse=0.198994
Model saved in file: /home/morbo/git/leela-zero/training/tf/leelaz-model-600000
Leela weights saved to /home/morbo/git/leela-zero/training/tf/leelaz-model-600000.txt
step 601000, policy=1.71325 mse=0.19333 reg=0.0999871 total=2.58656 (236.504 pos/s)
step 602000, policy=1.71425 mse=0.193062 reg=0.0999845 total=2.58648 (300.583 pos/s)
step 603000, policy=1.71125 mse=0.193567 reg=0.0999808 total=2.5855 (300.672 pos/s)
step 604000, policy=1.71831 mse=0.193841 reg=0.0999766 total=2.59365 (300.54 pos/s)
step 605000, policy=1.70846 mse=0.193841 reg=0.0999742 total=2.5838 (300.644 pos/s)
step 606000, policy=1.7172 mse=0.193675 reg=0.0999725 total=2.59187 (300.797 pos/s)
step 607000, policy=1.71105 mse=0.193599 reg=0.0999689 total=2.58542 (300.558 pos/s)
step 608000, policy=1.71839 mse=0.192411 reg=0.0999662 total=2.588 (300.499 pos/s)
step 608000, policy=1.70722 training accuracy=59.9941%, mse=0.201331
...
lowered learning rate
...
step 672000, policy=1.85378 training accuracy=57.1055%, mse=0.174458
step 680000, policy=1.67339 training accuracy=60.0908%, mse=0.218439
```

Note how both the policy and mse validation score suddenly get all over the place. And the errors start raising, not lowering...

**Comment by jkiliani** [comment]
This is the 192x10 bootstrap? You think it's getting worse, and can no longer be fairly compared to 128x20 due to the training code?

**Comment by jkiliani** [comment]
How often do you load new games into the training window with the bootstrap nets? Does this cause large jumps in MSE every time?

**Comment by jkiliani** [comment]
@Error323 stated in https://github.com/glinscott/leela-chess/issues/92#issuecomment-374364955 that the V2 code still does not shuffle the validation test set properly, could this account for your observations? He already pushed a fix for that...

**Comment by Error323** [comment]
Unless you're sweeping through the entire test set each validation step you want to increase the shuffle_size and sample size indeed.

**Comment by gcp** [comment]
Thanks. So this means that the training probably works correctly, but the validation is busted.

Increase of policy and MSE scores can be due to new training data.

**Comment by Error323** [comment]
That's what I would conclude :)

Op 20 mrt. 2018 12:49 p.m. schreef "Gian-Carlo Pascutto" <
notifications@github.com>:

**Comment by jkiliani** [comment]
Can you tell in spite of the obvious problem with the current validation code which of the two bootstraps currently fits the training data better?

**Comment by gcp** [comment]
Not without fixing them, no.
